core/shared/src/main/scala/zio/stm/ZSTM.scala

```diff
@@ -933,17 +933,26 @@ object ZSTM {
    * Atomically performs a batch of operations in a single transaction.
    */
   def atomically[R, E, A](stm: ZSTM[R, E, A]): ZIO[R, E, A] =
-    ZIO.accessM[R] { r =>
-      ZIO.effectSuspendTotalWith { (platform, fiberId) =>
-        tryCommit(platform, fiberId, stm, r) match {
-          case TryCommit.Done(io) => io // TODO: Interruptible in Suspend
-          case TryCommit.Suspend(journal) =>
-            val txnId     = makeTxnId()
-            val done      = new AtomicBoolean(false)
-            val interrupt = UIO(Sync(done)(done.set(true)))
-            val async     = ZIO.effectAsync(tryCommitAsync(journal, platform, fiberId, stm, txnId, done, r))
-
-            async ensuring interrupt
+    ZIO.uninterruptibleMask { restore =>
```

Comment:
Unfortunately this will decrease performance significantly. 

Could it be the issue is caused by the TODO on line 941? In which case a simpler fix might suffice.

Reply:
I think that is right but I'm not sure how we can avoid it. To avoid using `uninterruptibleMask` in the synchronous case I think we need to avoid performing any continuation after we call `tryCommit` in the synchronous case. But we already need to return an effect in the asynchronous branch and even in the synchronous branch if we change the result of `TryCommit` to return an `Exit` value we are still going to need a continuation to flatten the result back to a `ZIO` value.